protocols/request-response/tests/ping.rs

```diff
@@ -156,154 +160,127 @@ fn ping_protocol() {
-    let ping_proto2 = Behaviour::new(PingCodec(), protocols, cfg);
-    let mut swarm2 = Swarm::without_executor(trans, ping_proto2, peer2_id);
-
-    let addr = "/ip4/127.0.0.1/tcp/0".parse().unwrap();
-    swarm1.listen_on(addr).unwrap();
-
-    futures::executor::block_on(async move {
-        while swarm1.next().now_or_never().is_some() {}
-        let addr1 = Swarm::listeners(&swarm1).next().unwrap();
-
-        swarm2.behaviour_mut().add_address(&peer1_id, addr1.clone());
-        swarm2.behaviour_mut().send_request(&peer1_id, ping.clone());
-
-        // Wait for swarm 1 to receive request by swarm 2.
-        let event = loop {
-            futures::select!(
-                event = swarm1.select_next_some() => {
-                    if let SwarmEvent::Behaviour(Event::Message {
-                        peer,
-                        message: Message::Request { request, channel, .. }
-                    }) = event {
-                        assert_eq!(&request, &ping);
-                        assert_eq!(&peer, &peer2_id);
-
-                        drop(channel);
-                        continue;
-                    }
-                },
-                event = swarm2.select_next_some() => {
-                    if let SwarmEvent::Behaviour(ev) = event {
-                        break ev;
-                    }
-                },
-            )
-        };
-
-        let error = match event {
-            Event::OutboundFailure { error, .. } => error,
-            e => panic!("unexpected event from peer 2: {e:?}"),
-        };
+    let cfg = request_response::Config::default();
 
-        assert_eq!(error, OutboundFailure::ConnectionClosed);
+    let mut swarm1 = Swarm::new_ephemeral(|_| {
+        request_response::Behaviour::new(PingCodec(), protocols.clone(), cfg.clone())
     });
-}
+    let peer1_id = *swarm1.local_peer_id();
+    let mut swarm2 =
+        Swarm::new_ephemeral(|_| request_response::Behaviour::new(PingCodec(), protocols, cfg));
+    let peer2_id = *swarm2.local_peer_id();
+
+    swarm1.listen().await;
+    swarm2.connect(&mut swarm1).await;
+
+    swarm2.behaviour_mut().send_request(&peer1_id, ping.clone());
+
+    // Wait for swarm 1 to receive request by swarm 2.
+    let event = loop {
+        futures::select!(
+            event = swarm1.select_next_some() => {
+                if let SwarmEvent::Behaviour(request_response::Event::Message {
+                    peer,
+                    message: request_response::Message::Request { request, channel, .. }
+                }) = event {
+                    assert_eq!(&request, &ping);
+                    assert_eq!(&peer, &peer2_id);
+
+                    drop(channel);
+                    continue;
+                }
+            },
+            event = swarm2.select_next_some() => {
+                if let SwarmEvent::Behaviour(ev) = event {
+                    break ev;
+                }
+            },
+        )
+    };
```

Comment:
Can this not be done via the new shiny `drive` function?

Reply:
Probably, I've introduced that one only at the very end!

Reply:
I am not sure actually, this one is a bit different because the two peers operate autonomously. `drive` is useful if you want to "step" through a test in a more linear fashion because it gives you access to the `Swarm`s inbetween. Here, we can just create to tasks that operate independently.